v-next/hardhat-utils/src/eth.ts

```diff
@@ -87,7 +127,56 @@ export async function generateAddressBytes(): Promise<Uint8Array> {
  *
  * @returns A pseudo-random address.
  */
-export async function randomAddress(): Promise<PrefixedHexString> {
+export async function randomAddress(): Promise<Address> {
   const addressBytes = await generateAddressBytes();
   return bytesToHexString(addressBytes);
 }
+
+interface Artifact {
+  bytecode: string;
+  linkReferences: {
+    [sourceName: string]: {
+      [libraryName: string]: Array<{ start: number; length: number }>;
+    };
+  };
+}
+
+/**
+ * Links the bytecode of a contract with the provided library addresses.
+ *
+ * This function replaces placeholders in the bytecode with the actual
+ * addresses of the libraries. It throws an error if the link references
+ * for a library are undefined.
+ *
+ * @param artifact The artifact containing the bytecode and link references.
+ * @param libraries An array of libraries with their source names, contract
+ * names, and addresses.
+ * @returns The linked bytecode as a prefixed hexadecimal string.
+ * @throws LinkReferenceError If the link references for a library are undefined.
+ */
+export function linkBytecode(
+  artifact: Artifact,
+  libraries: Library[],
```

Comment:
I'm pretty sure that we accept bare names everywhere, so this is too restrictive.

In general we should do this:

- If we get a bare contract name, and it's unique, we accept it.
- If we accept a FQN (either formatted or as sourceName+contractName), we accept it.

And I'm thinking to add a third case:

- If we accept a unique suffix of a sourceName and a contractName, we accept it.

The reason for this later rule is that in the new build system, some source names are complex, like `/npm/@OpenZeppelin@1.2.3/contracts/token/Fool.sol`, and maybe `token/Foo.sol` is enough to identify it.

I have not decided about adding it yet. Just sharing it with you to get your opinion.